[Merged by Bors] - Variable size PoST

noamnelke opened the pull request:
## Motivation
https://github.com/spacemeshos/SMIPS/issues/6

## Changes
- Move space declaration from nipst to ATX header
- Return weights from active set calculation
- Use weights for block eligibility

## Todo
- [x] Implement efficient calculation of binomial cumulative distribution function using fixed decimal point
- [x] Implement weighted hare eligibility using the binomial CDF
- [ ] Update documentation as needed

## Test Plan
- [ ] Add unit tests
- [ ] Add system test for network with heterogeneously weighted miners

noamnelke committed:
move space from nipst to atx header

noamnelke committed:
accumulate epoch weight

noamnelke committed:
return weights from active set calculation

noamnelke committed:
use weight for eligibility

noamnelke committed:
use rand ports in api test

noamnelke committed:
lint fix

noamnelke committed:
Merge remote-tracking branch 'origin/develop' into var_post

noamnelke committed:
Merge remote-tracking branch 'origin/develop' into var_post

noamnelke committed:
hare weighted eligibility

noamnelke committed:
hare weighted voting

noamnelke commented:
bors try

noamnelke committed:
Merge remote-tracking branch 'origin/develop' into var_post

noamnelke committed:
restore protobuf dep

noamnelke committed:
set smaller space until we implement dividing by the space unit

noamnelke committed:
Merge remote-tracking branch 'origin/develop' into var_post

noamnelke commented:
bors try

noamnelke committed:
fix app test

noamnelke commented:
bors try

noamnelke commented:
bors try

noamnelke committed:
Merge remote-tracking branch 'origin/develop' into var_post

noamnelke commented:
bors try

noamnelke committed:
update fixed lib

noamnelke committed:
reafctor hare eligibility check

noamnelke committed:
improve should participate log

noamnelke committed:
Merge remote-tracking branch 'origin/develop' into var_post

noamnelke committed:
merge fixes

noamnelke committed:
Merge remote-tracking branch 'origin/develop' into var_post

noamnelke committed:
merge fixes

noamnelke committed:
revert accidentally removing dependency

noamnelke commented:
bors try

noamnelke committed:
fix app test, improve logging and fix minor issues

noamnelke commented:
bors try

noamnelke commented:
bors try

noamnelke committed:
use ed key for hare eligibility weight lookup

noamnelke commented:
bors try

noamnelke commented:
bors try

noamnelke committed:
Merge remote-tracking branch 'origin/develop' into var_post

noamnelke committed:
reorder tests in travis from long to short

noamnelke commented:
bors try

noamnelke commented:
bors try

noamnelke committed:
reinstate sync test

noamnelke committed:
improve logging

noamnelke committed:
fix sync automation test

noamnelke committed:
Merge remote-tracking branch 'origin/develop' into var_post

noamnelke commented:
bors try

noamnelke commented:
bors try

noamnelke commented:
bors try

noamnelke commented:
bors try-

noamnelke commented:
bors try

noamnelke commented:
While the last `bors try` failed, the failure was random (confirmed by running locally and it passed). I tried to re-run but for some reason Github Actions doesn't get triggered by the push to `trying`. Need to figure this out, but it's not code-related.

lrettig commented:
bors try

lrettig commented:
bors try-

lrettig commented:
bors try

lrettig commented:
bors try

noamnelke committed:
make test default ports different than testnet ports

noamnelke committed:
change hare committee size to 800 everywhere

noamnelke commented:
bors try

noamnelke commented:
bors try-

noamnelke commented:
bors try

noamnelke committed:
missed a spot

noamnelke commented:
bors try

noamnelke commented:
bors try-

noamnelke commented:
bors try

noamnelke committed:
move multi_node_test so that it uses config from correct path

noamnelke commented:
bors try-

noamnelke commented:
bors try

noamnelke committed:
use small hare committee because of non-weighted oracle

noamnelke commented:
bors try

noamnelke committed:
fix more test ports

noamnelke commented:
bors try

noamnelke committed:
heterogeneous miner wieghts in app test

moshababo commented on activation/atxdb.go:
```diff
@@ -201,129 +205,110 @@ func (db *DB) createTraversalActiveSetCounterFunc(countedAtxs map[string]types.A
+			return totalWeight, nil
 		}
-		// if not found, keep running mutex and calculate active set size
+		// if not found, keep running mutex, ensure it's still in the pending map and calculate total weight
```
What is the scenario in which a hash bearing a mutex-lock was found in `db.pendingTotalWeight`, but once unlocked it was not found in `totalWeightCache`? Is it only if it was cleared out of the cache immediately after it was put there?

noamnelke replied:
If there was an error calculating it:
https://github.com/spacemeshos/go-spacemesh/blob/a44a9c80e4a94621f97992f35f3d67ab24450bca/activation/atxdb.go#L289-L294

moshababo commented on activation/atxdb.go:
```diff
@@ -55,34 +59,35 @@ type DB struct {
+	log                 log.Log
+	calcTotalWeightFunc func(targetEpoch types.EpochID, blocks map[types.BlockID]struct{}) (map[string]uint64, error)
+	processAtxMutex     sync.Mutex
+	epochWeightMutex    sync.Mutex // used only for updating total epoch weight, reading can be done lock-free
```
You have different suffixes for `sync.Mutex` fields: `Mutex` & `Lock`. You should stick with one (I prefer `Mtx`).

moshababo commented on activation/atxdb.go:
```diff
@@ -55,34 +59,35 @@ type DB struct {
+	LayersPerEpoch      uint16
+	nipstValidator      nipstValidator
+	pendingTotalWeight  map[types.Hash12]*sync.Mutex
+	pTotalWeightLock    sync.Mutex
```
Although longer, it would be easier to understand what this lock is associated with if called `pendingTotalWeightMtx`.

moshababo commented on activation/atxdb.go:
```diff
@@ -201,129 +205,110 @@ func (db *DB) createTraversalActiveSetCounterFunc(countedAtxs map[string]types.A
+	return totalWeight, nil
 }
 
 func (db *DB) deleteLock(viewHash types.Hash12) {
```
Perhaps also change the name of this function, it's quite confusing.

moshababo commented on activation/atxdb.go:
```diff
@@ -659,6 +648,37 @@ func (db *DB) GetPosAtxID() (types.ATXID, error) {
+	}
+	newWeight := weight + additionalWeight
+	if newWeight < weight {
+		newWeight -= newWeight + 1 // this under-flows newWeight so it holds the max possible value
```
Why not simply assign `math.MaxUint64`? It will be much easier to understand.

moshababo replied:
You are basically reducing the total weight here (which was larger than the max possible). Wouldn't it create discrepancies in the eligibility proofs? It's not clear if the current implementation is equipped to handle this scenario.

moshababo commented on activation/validator.go:
```diff
@@ -34,11 +34,11 @@ func (v *Validator) Validate(minerID signing.PublicKey, nipst *types.NIPST, expe
 
-	if nipst.Space < v.postCfg.SpacePerUnit {
-		return fmt.Errorf("PoST space (%d) is less than a single space unit (%d)", nipst.Space, v.postCfg.SpacePerUnit)
+	if space < v.postCfg.SpacePerUnit {
```
`space` validation should probably move to `SyntacticallyValidateAtx`.

noamnelke replied:
Not sure about this. The `space` still needs to be passed into this method, because we need it in order to validate the PoST. The validator has access to the post config, which the ATX DB doesn't, so validating it here is much easier.

moshababo commented on activation/validator.go:
```diff
@@ -34,11 +34,11 @@ func (v *Validator) Validate(minerID signing.PublicKey, nipst *types.NIPST, expe
 
-	if nipst.Space < v.postCfg.SpacePerUnit {
-		return fmt.Errorf("PoST space (%d) is less than a single space unit (%d)", nipst.Space, v.postCfg.SpacePerUnit)
+	if space < v.postCfg.SpacePerUnit {
```
What's the reason for this validation? 
If it's about having a min `space` value, please describe the reasons for such kind of limitation, and in addition upgrade the config param name (it doesn't need to come from the `post` repo config. If it won't be streightforward to implement, leave a TODO and i'll fix it in the upcoming merge with https://github.com/spacemeshos/go-spacemesh/pull/2149).

noamnelke replied:
This implementation is not complete - ultimately we want to express the space in the number of units (so if `SpacePerUnit` is 100GB then you can only commit in increments of that). Then the validation will just be that it's `>0`.

I agree that it shouldn't be in the `post` repo config, not sure how it got there in the first place. Removing it from there won't be trivial - it's used in that package, maybe not in the same way I use it here.

I'll add a TODO.

moshababo commented on common/types/activation.go:
```diff
@@ -280,10 +292,6 @@ type PoetRound struct {
 type NIPST struct {
-	// space is the amount of storage which the prover
-	// requires to dedicate for generating the NIPST.
-	Space uint64
```
What's the reason for moving this to the ATX header?

noamnelke replied:
The ATX header contains all the fields that need to be used outside of validation, we use the body only to validate the ATX and then pass around the header. This means we don't need to load the ATX body from disk or keep it in memory.

The changes in this PR require us to access the space outside of validation (e.g. to summarize total ATX weight) so it makes sense to move the space into the header.

moshababo commented on api/config/config.go:
```diff
@@ -7,13 +7,13 @@ import (
-	defaultGRPCServerPort          = 9091
-	defaultNewGRPCServerPort       = 9092
+	defaultGRPCServerPort          = 19091
+	defaultNewGRPCServerPort       = 19092
```
What's for reason for this change?

noamnelke replied:
I had port conflicts with my local node when running tests locally. But you're right that changing the ports isn't the solution. I've now committed another change that reverts this change and, instead, introduces the concept of `DefaultTestConfig` where the ports are offset.

moshababo commented on common/types/activation.go:
```diff
@@ -111,6 +112,12 @@ func (atxh *ActivationTxHeader) SetID(id *ATXID) {
 
+// GetWeight returns the ATX's weight = space * ticks.
+func (atxh *ActivationTxHeader) GetWeight() uint64 {
+	// TODO: Limit the number of bits this can occupy
```
Why is this unimplemented? It's the basis for a few other overflow checks.

noamnelke replied:
We still don't know how PoET ticks are going to be implemented/normalized.

I can do a check and panic on overflow, but what I meant with this TODO was higher level -- design the parts of the weight in a way that an overflow would be unlikely. This one is only for a single ATX, preventing overflows when summing up an epoch's total weight is even more challenging.

moshababo commented on hare/eligibility/oracle.go:
```diff
@@ -166,69 +166,166 @@ func (o *Oracle) buildVRFMessage(layer types.LayerID, round int32) ([]byte, erro
+	actives, err := o.actives(layer)
+	if err != nil {
+		if err == errGenesis { // we are in genesis
+			return o.genesisTotalWeight / 50, nil
```
What's the 50 is about? Are we expected to have this fixed number of miners?

moshababo commented on activation/atxdb.go:
```diff
@@ -201,129 +205,110 @@ func (db *DB) createTraversalActiveSetCounterFunc(countedAtxs map[string]types.A
-	mu.Unlock()
+	totalWeight = 0
+	for _, w := range atxWeights {
+		totalWeight += w
```
Potential uint64 overflow?

moshababo commented on activation/atxdb.go:
```diff
@@ -201,129 +205,110 @@ func (db *DB) createTraversalActiveSetCounterFunc(countedAtxs map[string]types.A
-		return 0, fmt.Errorf("publication epoch cannot be less than 1, found %v", pubEpoch)
+// CalcTotalWeightFromView traverses the provided view and returns the total weight of the ATXs found targeting
+// targetEpoch. The function returns error if the view is not found.
+func (db *DB) CalcTotalWeightFromView(view []types.BlockID, targetEpoch types.EpochID) (uint64, error) {
```
What's the purpose of this? Looks like it's not used in non-test code.

antonlerner replied:
Do we need to traverse a view to calc miner weight? currently we do not include a view in ATX and just declare the list of ATXs as an active set... where does this traversing happens?

moshababo commented on hare/eligibility/oracle.go:
```diff
@@ -166,69 +166,166 @@ func (o *Oracle) buildVRFMessage(layer types.LayerID, round int32) ([]byte, erro
+	return fixed.FracFromBytes(sha[:8])
+}
+
+func (o *Oracle) prepareEligibilityCheck(layer types.LayerID, round int32, committeeSize int, id types.NodeID, sig []byte) (fixed.Fixed, fixed.Fixed, fixed.Fixed, bool, error) {
```
Consider returning a struct with documented fields instead of the 3 unnamed `fixed.Fixed`.

moshababo commented on hare/eligibility/oracle.go:
```diff
@@ -166,69 +166,166 @@ func (o *Oracle) buildVRFMessage(layer types.LayerID, round int32) ([]byte, erro
+
+	// ensure miner weight fits in int
+	if uint64(int(minerWeight)) != minerWeight {
+		panic(fmt.Sprintf("minerWeight overflows int (%d)", minerWeight))
```
Looks like it would be better to panic when `minerWeight` was calculated. 
But regardless of that, what's the greater solution besides panicking? 
Is there any other solution besides defining the max allowed bits (64 or more) and define/validate max allowed space+ticks combination and number of miners (for the total calc)?

moshababo commented on hare/eligibility/oracle.go:
```diff
@@ -166,69 +166,166 @@ func (o *Oracle) buildVRFMessage(layer types.LayerID, round int32) ([]byte, erro
+	if uint64(int(minerWeight)) != minerWeight {
+		panic(fmt.Sprintf("minerWeight overflows int (%d)", minerWeight))
+	}
+	n := fixed.New(int(minerWeight))
```
Why `Fixed` doesn't work with uint64?

moshababo commented on hare/eligibility/oracle.go:
```diff
@@ -166,69 +166,166 @@ func (o *Oracle) buildVRFMessage(layer types.LayerID, round int32) ([]byte, erro
+
+	// calc p
+	p := fixed.One
+	if committeeSize < int(totalWeight) {
```
Why not upgrading `committeeSize` to uint64 instead of downgrading `totalWeight` to int, and potentially overflow?

antonlerner commented on api/config/config.go:
```diff
@@ -7,13 +7,13 @@ import (
 	defaultStartGRPCServer         = false
-	defaultGRPCServerPort          = 9091
-	defaultNewGRPCServerPort       = 9092
+	defaultGRPCServerPort          = 19091
```
does this needs to go in the main branch? I think we currently rely on old ports for wallet app and such...

antonlerner replied:
anyways i think this needs to be in a separate pr

antonlerner commented on cmd/hare/oracle.go:
```diff
@@ -16,13 +16,21 @@ func (bo *hareOracle) IsIdentityActiveOnConsensusView(string, types.LayerID) (bo
-	// note: we don't use the proof in the oracle server. we keep it just for the future syntax
-	// todo: maybe replace k to be uint32 like hare wants, and don't use -1 for blocks
+func (bo *hareOracle) Validate(layer types.LayerID, round int32, committeeSize int, id types.NodeID, sig []byte, eligibilityCount uint16) (bool, error) {
+	if eligibilityCount != 1 {
```
add comment here

antonlerner commented on cmd/hare/oracle.go:
```diff
@@ -16,13 +16,21 @@ func (bo *hareOracle) IsIdentityActiveOnConsensusView(string, types.LayerID) (bo
 	return bo.oc.Eligible(layer, round, committeeSize, id, sig)
 }
 
+func (bo *hareOracle) CalcEligibility(layer types.LayerID, round int32, committeeSize int, id types.NodeID, sig []byte) (uint16, error) {
```
why return integer if only values are 0 or 1?

antonlerner commented on config.toml:
```diff
@@ -59,8 +59,8 @@ default-timeout-latency = "10s"
 hare-round-duration-sec = "5"
-hare-committee-size = 10
-hare-max-adversaries = 5
+hare-committee-size = 800
```
what does this mean now? is this total weight? how many miners does this represent? please comment

antonlerner commented on hare/eligibility/oracle.go:
```diff
@@ -166,69 +166,166 @@ func (o *Oracle) buildVRFMessage(layer types.LayerID, round int32) ([]byte, erro
+				log.Uint64("n", uint64(n.Value())),
+				log.Uint64("p", uint64(p.Value())),
+			)
+			panic(msg)
```
a panic within recover from panic?

noamnelke committed:
Merge remote-tracking branch 'origin/develop' into var_post

noamnelke committed:
post merge fixes

noamnelke committed:
different ports in test config

noamnelke committed:
clarify post space config

noamnelke committed:
Merge remote-tracking branch 'origin/develop' into var_post

noamnelke committed:
go mod tidy

noamnelke committed:
lint fix

noamnelke committed:
Merge remote-tracking branch 'origin/develop' into var_post

noamnelke committed:
revert changing post param name

noamnelke committed:
fix issues in mining test

noamnelke committed:
Merge remote-tracking branch 'origin/develop' into var_post

noamnelke committed:
add benchmark and improve logging

sudachen committed:
wighted hare consensus fixes

noamnelke committed:
cleanup old post code

noamnelke committed:
Merge remote-tracking branch 'origin/develop' into var_post

noamnelke committed:
merge fixes

noamnelke committed:
Merge remote-tracking branch 'origin/develop' into var_post

noamnelke committed:
fix merge-induced deadlock

noamnelke committed:
revert accidental test change

noamnelke committed:
another merge screwup

noamnelke committed:
use list of ATXs to compute epoch weight

noamnelke committed:
Merge remote-tracking branch 'origin/develop' into var_post

noamnelke committed:
fix after merge

noamnelke commented:
bors try

noamnelke committed:
another fix after merge

noamnelke commented:
bors try

noamnelke committed:
make hare assertions consider eligibilities

noamnelke commented:
bors try

noamnelke committed:
update fixed lib and reduce input size

noamnelke commented:
bors try

noamnelke committed:
go mod tidy

noamnelke commented:
bors try

noamnelke committed:
normalize hare elig. params when total storage is less units than committee size

noamnelke commented:
bors try

noamnelke committed:
fallback positioning atx to golden atx

noamnelke committed:
lint fix

noamnelke commented:
bors try

noamnelke committed:
handle genesis with miners != 50

noamnelke committed:
Merge remote-tracking branch 'origin/develop' into var_post

noamnelke commented:
bors try

noamnelke committed:
fix bad config

noamnelke committed:
increase dump timeout

noamnelke committed:
Merge remote-tracking branch 'origin/develop' into var_post

noamnelke committed:
merge fixes

noamnelke committed:
moar merge fixes

noamnelke commented:
bors try

noamnelke committed:
temp skip sync unit test

noamnelke commented:
bors try

noamnelke commented:
bors try

noamnelke commented:
bors retry

noamnelke commented:
bors try-

noamnelke commented:
bors try

noamnelke committed:
Merge remote-tracking branch 'origin/develop' into var_post

noamnelke committed:
improve log

noamnelke commented:
bors try

noamnelke committed:
log cleanup post merge

noamnelke committed:
try new test config

noamnelke commented:
bors try

noamnelke committed:
change vrf from bls to ed

noamnelke committed:
allow running poet harness concurrently

noamnelke committed:
Merge remote-tracking branch 'origin/develop' into var_post

noamnelke committed:
unflake test

noamnelke commented:
bors try

noamnelke committed:
lint fix

noamnelke committed:
Revert "allow running poet harness concurrently"

This reverts commit 5e6f96ba71fdbe3af7edb5a7453a943ca5c39c61.

noamnelke commented:
bors try

noamnelke committed:
add tests and remove error from vrf signer and verifier

noamnelke committed:
oopsie

noamnelke committed:
testfix

noamnelke committed:
skip block eligibility check in genesis

noamnelke committed:
Merge remote-tracking branch 'origin/develop' into var_post

lrettig committed:
Workarounds for concurrency issues in tests

Add brief timeouts to make sure other goroutines that the tests rely on
have started

noamnelke committed:
make assertion consider layers per epoch

noamnelke committed:
add space-to-commit to all configs

noamnelke committed:
improve log

noamnelke committed:
Merge remote-tracking branch 'origin/develop' into var_post

noamnelke commented:
bors try

noamnelke committed:
consistent serialization + remove redundant hashing

antonlerner committed:
fixed sync start after enough peers were received

antonlerner committed:
fixed random seed causing flaky tests in Test_VrfSignVerify

antonlerner commented:
bors try

antonlerner committed:
fixed is synced merge error

antonlerner commented:
bors try

antonlerner committed:
bugfix

antonlerner commented:
bors try

antonlerner committed:
after fmt

antonlerner commented:
bors try

antonlerner commented:
bors try

antonlerner commented:
bors try

noamnelke committed:
aggregate errors in ContextuallyValidBlock

noamnelke commented:
bors try

antonlerner committed:
fixed syncer start test flakyness

antonlerner committed:
Merge branch 'var_post' of https://github.com/spacemeshos/go-spacemesh into var_post

antonlerner committed:
added ttl to fetch requests and patched poll layer in weakly synced

antonlerner commented:
bors try

antonlerner committed:
added more retries per request

antonlerner commented:
bors try

antonlerner commented:
bors try

antonlerner committed:
Merge branch 'develop' into var_post

antonlerner commented:
bors try

antonlerner commented:
bors try

antonlerner committed:
added prints to find clock leak

antonlerner committed:
Merge branch 'var_post' of https://github.com/spacemeshos/go-spacemesh into var_post

antonlerner committed:
added more time to MultipleCps test

antonlerner commented:
bors try

antonlerner committed:
some defensive code to handle unit test network races in sync

antonlerner commented:
bors try

antonlerner committed:
patched another race in sync tests

noamnelke commented on layerfetcher/layers.go:
```diff
@@ -261,7 +265,16 @@ func (l *Logic) PollLayer(ctx context.Context, layer types.LayerID) chan LayerPr
+		return
+	}
+
+	l.log.Error("got")
 	l.layerHashResM.Lock()
+	l.log.Error("got here")
```
looks like this should be cleaned up @antonlerner

antonlerner committed:
test will not print anyways now, removed debug prints

antonlerner commented:
bors try

antonlerner committed:
removed more unnecesary prints

antonlerner commented:
bors try

antonlerner commented:
bors try

antonlerner committed:
fixed sync test multi threading hickups fail

noamnelke commented:
bors merge

noamnelke commented:
bors merge
